Paper: Training machine learning models faster with PyTorch and Dask

stsievert opened the pull request:
* **Title:** Training machine learning models faster with Dask
* **Authors:** @JoeHolt and I
* **Presentation type:** Poster.

The **current draft** of this paper can be found at https://procbuild.scipy.org/download/stsievert-adadamp. If it isn't built, visit http://procbuild.scipy.org/ and trigger a build. and the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper) has more information.

*Editor*: <!--editor-->@cbcunc<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@richardliaw<!--end-reviewers-list-->

stsievert committed:
Initial commit

stsievert committed:
Add figs

stsievert committed:
edit title

stsievert committed:
edits

stsievert committed:
add img

stsievert committed:
incorporate joes comments

deniederhut commented:
@scoobies assign @cbcunc as editor

scoobies commented:
Assigned! @cbcunc is now the editor

cbcunc commented:
Editor’s initial quickie look-over:

- [x] Make sure the paper builds on procbuild
- [x] Make sure that the sections are filled in
- [x] Make sure 8 pages or less exclusive of references
- [x] No changes outside paper directory

deniederhut commented:
@scoobies add @richardliaw to reviewers

scoobies commented:
@richardliaw added to the reviewers list!

deniederhut commented:
We've been having difficulty finding reviewers for some papers, so to make sure everyone has enough time to do their reviews, we will be bumping the deadline for the initial review back to June 17th.

deniederhut commented:
Hey @richardliaw! Today is the deadline for initial reviews. Have you had a chance to review this paper?

richardliaw commented on papers/scott_sievert/main.rst:
```diff
@@ -0,0 +1,726 @@
+   balles2016coupling, byrd2012}.    Of course, these methods are adaptive so
+   computing the batch size requires computation (though there are
+   workarounds~\cite{sievert2021improving, balles2016coupling}). The
+   convergence results for these adaptive methods suggest passive methods of
+   increasing the batch size \cite{sievert2021improving}.
```
Hmm, I didn't quite understand this sentence. Should this sentence be formed as "[sic] These results suggest *that* passive methods ... are better" (or something?)

stsievert replied:
Resolved. Thanks!

(I deleted this sentence – turns out the passive methods aren't relevant).

richardliaw commented on papers/scott_sievert/main.rst:
```diff
@@ -0,0 +1,726 @@
+
+.. latex::
+
+   The key feature of AdaDamp is that the number of workers grows
+   with the batch size. Then, the model update time is agnostic to the batch size
+   (provided communication is instantaneous). This has been shown empirically:
+   Goyal et al. grow the batch size (and the number of workers with it) by a
+   factor of $44$ but the time for a single model update only increases by a
+   factor of $1.13$~\cite[Sec.~5.5]{goyal2017accurate}.
```
Maybe you want to highlight this key feature in the intro. I read up until here and didn't realize this is what the paper was about until this sentence.

stsievert replied:
Good idea. I mentioned it in the contributions section, just below the bullet points.

richardliaw commented on papers/scott_sievert/main.rst:
```diff
@@ -0,0 +1,726 @@
+   factor of $44$ but the time for a single model update only increases by a
+   factor of $1.13$~\cite[Sec.~5.5]{goyal2017accurate}.
+
+In distributed systems, this behavior—growing the amount of data with the
+number of workers—is known as "weak scaling." By contrast, "strong scaling" has
+a fixed batch size and treats the number of workers as an internal detail. Of
+course, weak scaling exhibits better scaling than strong scaling
+:cite:`qi2017paleo`.
```
Should this go in the intro too? It reads more like context.

richardliaw commented on papers/scott_sievert/main.rst:
```diff
@@ -0,0 +1,726 @@
+cluster. This package reduces the amount of time required to train a model with
+the current centralized setup. However, it can be further accelerated by
+integration with PyTorch's distributed communication package as illustrated by
+extensive simulations. In summary, the expected gains are to go from training
+requiring about 120 minutes to 45 minutes.
```
maybe mention that this is for a specific model? 120 minutes can be considered very long or very short -- it's model dependent.

richardliaw commented on papers/scott_sievert/main.rst:
```diff
@@ -0,0 +1,726 @@
+computation is bottleneck with this model/dataset/hardware.
+
+.. figure:: figs/decentralized-moderate/training_time.png
+   :align: center
```
do you incorporate any graphs regarding time vs loss in the manuscript?

I saw that there was a `epoch_score__loss/acc.pdf`, but it didn't seem to be included in the manuscript.

stsievert replied:
The graph with loss mirrors the one with accuracy; I've included them in the appendix.

richardliaw reviewed (commented):
Overall nice work! Tried to avoid nit-picking.

cbcunc commented:
@scoobies help

scoobies commented:
Hello @cbcunc, here are the things you can ask me to do:

```

# List all available commands
@scoobies help

# Show our community Code of Conduct and Guidelines
@scoobies code of conduct

# Add to this issue's reviewers list
@scoobies add @username to reviewers

# Remove from this issue's reviewers list
@scoobies remove @username from reviewers

# Assign a user as the editor of this submission
@scoobies assign @username as editor

# Remove the editor assigned to this submission
@scoobies remove editor

# Add a user to this issue's assignees list
@scoobies add assignee: @username

# Remove a user from this issue's assignees list
@scoobies remove assignee: @username

# Builds paper
@scoobies build paper

# Checks build status
@scoobies build status

# Check the references of the paper for missing DOIs
# Optionally, it can be run on a non-default branch 
@scoobies check references
@scoobies check references from custom-branch-name

# Label issue with: paper
@scoobies mark paper

# Label issue with: needs-review. Remove labels: unready, needs-more-review, pending-comment, ready
@scoobies mark needs review

# Label issue with: needs-more-review. Remove labels: unready, needs-review, pending-comment, ready
@scoobies mark needs more review

# Label issue with: pending-comment. Remove labels: unready, needs-review, needs-more-review, ready
@scoobies mark pending comment

# Label issue with: ready. Remove labels: unready, needs-review, needs-more-review, pending-comment
@scoobies mark ready

# Label issue with: unready. Remove labels: needs-review, needs-more-review, pending-comment, ready
@scoobies mark not ready

# Label issue with: does_not_build:server
@scoobies mark server fail

# Remove labels: does_not_build:server
@scoobies mark server success

```

cbcunc commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- 10.1007/s10107-012-0572-5 may be a valid DOI for title: Sample size selection in optimization methods for machine learning
- 10.1109/cvpr.2019.01138 may be a valid DOI for title: A sufficient condition for convergences of adam and rmsprop
- 10.1145/2640087.2644155 may be a valid DOI for title: Scaling Distributed Machine Learning with the Parameter Server.
- 10.1109/icdm.2016.0074 may be a valid DOI for title: HogWild++: A New Mechanism for Decentralized Asynchronous Stochastic Gradient Descent
- 10.1007/978-3-319-46128-1_50 may be a valid DOI for title: Linear convergence of gradient and proximal-gradient methods under the polyak-łojasiewicz condition

INVALID DOIs

- None
```

cbcunc commented:
@stsievert See the DOI recommendations above at https://github.com/scipy-conference/scipy_proceedings/pull/614#issuecomment-866423273 to see if it helps you to flesh out your references. We need each reference to have a DOI wherever possible. Thanks.

deniederhut commented:
@scoobies mark needs more review

deniederhut commented:
@stsievert to give everyone more time to respond to their reviews, we've pushed back the deadline for authors' final revisions to July 1st

stsievert committed:
add and change figures

stsievert committed:
edits

cbcunc commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- 10.1007/s10107-012-0572-5 may be a valid DOI for title: Sample size selection in optimization methods for machine learning
- 10.1109/cvpr.2019.01138 may be a valid DOI for title: A sufficient condition for convergences of adam and rmsprop
- 10.1145/2640087.2644155 may be a valid DOI for title: Scaling Distributed Machine Learning with the Parameter Server.
- 10.1109/icdm.2016.0074 may be a valid DOI for title: HogWild++: A New Mechanism for Decentralized Asynchronous Stochastic Gradient Descent
- 10.1007/978-3-319-46128-1_50 may be a valid DOI for title: Linear convergence of gradient and proximal-gradient methods under the polyak-łojasiewicz condition

INVALID DOIs

- None
```

stsievert committed:
add some dois

stsievert commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1137/16M1080173 is OK
- 10.1007/s10107-012-0572-5 is OK
- 10.1109/cvpr.2019.01138 is OK
- 10.1109/icdm.2016.0074 is OK
- 10.1007/978-3-319-46128-1_50 is OK

MISSING DOIs

- 10.1145/2640087.2644155 may be a valid DOI for title: Scaling Distributed Machine Learning with the Parameter Server

INVALID DOIs

- None
```

stsievert committed:
tighten up references

stsievert committed:
run through

stsievert committed:
edits

stsievert commented:
Thanks for the review @richardliaw! I've addressed your comment and done another pass over the paper.

I'd appreciate another review if you're willing and able. Don't be afraid to nitpick (especially given the mistakes I caught!). I have verified https://procbuild.scipy.org/download/stsievert-adadamp is up-to-date.

deniederhut commented:
Hey @richardliaw ! Do you feel that Scott's changes have addressed your comments?

deniederhut commented:
@scoobies mark pending comment

cbcunc commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1137/16M1080173 is OK
- 10.1007/s10107-012-0572-5 is OK
- 10.1137/0330046 is OK
- 10.1109/cvpr.2019.01138 is OK
- 10.3929/ethz-b-000319485 is OK
- 10.14778/3415478.3415530 is OK
- 10.1109/ICDM.2016.0074 is OK
- 10.1007/978-3-319-46128-1_50 is OK

MISSING DOIs

- 10.1145/2640087.2644155 may be a valid DOI for title: Scaling Distributed Machine Learning with the Parameter Server

INVALID DOIs

- None
```

cbcunc commented:
@scoobies mark ready
